### PR TITLE
Add a maxPrice option to the client retrieval CLI

### LINKED ISSUE #2151: Client Retrieve Command Should Specify Parameters For Deal
# Why

Currently, the retrieval command queries all the miners it knows who may have a payloadCID and simply retrieves from the first one it hears back from (offers[0] in the list of query offers.

This presents a number of issues -- most specifically, that the price for the retrieval could be anything!

# What

As a starting point, we should probably sort offers by price.

However, we should also specify a maximum price on the command line -- possibly as a mandatory parameter (possibly optional but with a default limit and a message if it goes over)

We could also optionally allow specifying other parameters as needed.

### PR DESCRIPTION
**Pull request by ingar**
## Summary
- Add a `maxPrice` option to the `client retrieve` CLI
- Use the lowest-priced offer if multiple retrieval offers are returned

Resolves #2151

### CODE REVIEW

File: cli/client.go [hunk 1]
```diff
@@ -584,6 +593,17 @@ var clientRetrieveCmd = &cli.Command{
 			return fmt.Errorf("The received offer errored: %s", offer.Err)
 		}
 
+		if cctx.String("maxPrice") != "" {
+			maxPrice, err := types.ParseFIL(cctx.String("maxPrice"))
+			if err != nil {
+				return err
```
Comment (magik6k) [inline cli/client.go:599]:
```suggestion
				return xerrors.Errorf("parsing maxPrice: %w", maxPrice)
```

File: cli/client.go [hunk 2]
```diff
@@ -584,6 +593,17 @@ var clientRetrieveCmd = &cli.Command{
...
+			}
+
+			if offer.MinPrice.GreaterThan(types.BigInt(maxPrice)) {
+				return xerrors.Errorf("Failed to find offer satisfying maxPrice: %w", maxPrice)
```
Comment (magik6k) [inline cli/client.go:603]: %w is for errors
```suggestion
				return xerrors.Errorf("failed to find offer satisfying maxPrice: %s", maxPrice)
```

File: cli/client.go [hunk 3]
```diff
@@ -584,6 +593,17 @@ var clientRetrieveCmd = &cli.Command{
 			return fmt.Errorf("The received offer errored: %s", offer.Err)
 		}
 
+		if cctx.String("maxPrice") != "" {
+			maxPrice, err := types.ParseFIL(cctx.String("maxPrice"))
```
Comment (magik6k) [inline cli/client.go:597]: I feel like we really should have some default (like 1FIL)

Comment (ingar) [reply cli/client.go:597]: Will do

File: cli/client.go [hunk 4]
```diff
@@ -593,15 +595,19 @@ var clientRetrieveCmd = &cli.Command{
...
+			maxPriceFil, err := types.ParseFIL(cctx.String("maxPrice"))
 			if err != nil {
-				return err
+				return xerrors.Errorf("parsing maxPrice: %w", err)
 			}
 
-			if offer.MinPrice.GreaterThan(types.BigInt(maxPrice)) {
-				return xerrors.Errorf("Failed to find offer satisfying maxPrice: %w", maxPrice)
-			}
+			maxPrice = types.BigInt(maxPriceFil)
```
Comment (magik6k) [inline cli/client.go:606]: Should be
```suggestion
			maxPrice = types.FromFil(maxPriceFil)
```

Otherwise this will set the max to 1 attoFil

Comment (magik6k) [reply cli/client.go:606]: Wait, meant here - https://github.com/filecoin-project/lotus/pull/2347/files#diff-b47a2d7fe9543751ffd6d9e6a5be7433R598

**Pull request merged**